File: packages/core/src/citable-path.ts
/**
 * Citable-path addressing for the document tree.
 *
 * A pin-cite (CONTEXT.md) resolves to a specific citable unit — the smallest
 * document-tree node a citation can address. A {@link CitablePath} is the
 * machine representation of that address: a `documentId` plus an ordered list
 * of typed segments from the document root down to the cited node
 * (Part → section → subsection → clause).
 *
 * Modelling the path as structured segments (rather than a display string like
 * "RTA s. 49(1)(a)") makes the hierarchical relation a clean prefix check and
 * keeps the cross-language conformance vectors free of any parsing dialect.
 */

import { z } from 'zod'

/** The node kinds a document tree distinguishes, root-first. */
export const SEGMENT_KINDS = ['document', 'part', 'section', 'subsection', 'clause'] as const

export type SegmentKind = (typeof SEGMENT_KINDS)[number]

/** One step in a citable path: the node's kind and its label (e.g. section "49"). */
export const citablePathSegmentSchema = z
  .object({
    kind: z.enum(SEGMENT_KINDS),
    label: z.string().min(1),
  })
  .strict()

export type CitablePathSegment = z.infer<typeof citablePathSegmentSchema>

/**
 * The address of one document-tree node: which document, and the ordered
 * segments from its root to the node. An empty `segments` list addresses the
 * document root itself.
 */
export const citablePathSchema = z
  .object({
    documentId: z.string().min(1),
    segments: z.array(citablePathSegmentSchema),
  })
  .strict()

export type CitablePath = z.infer<typeof citablePathSchema>

/** Validate and normalize an untyped value into a {@link CitablePath}. */
export function parseCitablePath(value: unknown): CitablePath {
  return citablePathSchema.parse(value)
}

function segmentsEqual(a: CitablePathSegment, b: CitablePathSegment): boolean {
  return a.kind === b.kind && a.label === b.label
}

/** True when two citable paths address the same node of the same document. */
export function citablePathsEqual(a: CitablePath, b: CitablePath): boolean {
  if (a.documentId !== b.documentId) return false
  if (a.segments.length !== b.segments.length) return false
  return a.segments.every((segment, index) => segmentsEqual(segment, b.segments[index]!))
}

/**
 * True when `maybeAncestor` is a strict prefix of `maybePath` within the same
 * document — i.e. `maybeAncestor` sits strictly higher on the same line.
 * Kinds must agree segment-for-segment, so a mislabeled coordinate (a clause
 * "1" vs a subsection "1") is not treated as the same node.
 */
export function isProperAncestor(maybeAncestor: CitablePath, maybePath: CitablePath): boolean {
  if (maybeAncestor.documentId !== maybePath.documentId) return false
  if (maybeAncestor.segments.length >= maybePath.segments.length) return false
  return maybeAncestor.segments.every((segment, index) =>
    segmentsEqual(segment, maybePath.segments[index]!),
  )
}

/** True when `maybeDescendant` sits strictly lower on the same line as `maybePath`. */
export function isProperDescendant(maybeDescendant: CitablePath, maybePath: CitablePath): boolean {
  return isProperAncestor(maybePath, maybeDescendant)
}

/** Render a citable path for human review (e.g. "RTA / Part V / s. 49 / (1) / (a)"). */
export function formatCitablePath(path: CitablePath): string {
  const head = path.documentId
  const tail = path.segments.map((segment) => {
    switch (segment.kind) {
      case 'part':
        return `Part ${segment.label}`
      case 'section':
        return `s. ${segment.label}`
      case 'subsection':
      case 'clause':
        return `(${segment.label})`
      case 'document':
        return segment.label
    }
  })
  return [head, ...tail].join(' / ')
}

File: evals/src/owners_manual_evals/citable_path.py
"""Citable-path addressing for the document tree (Python side).

Mirror of ``packages/core/src/citable-path.ts``. A pin-cite (CONTEXT.md)
resolves to a citable unit — the smallest document-tree node a citation can
address. A :class:`CitablePath` is that address: a ``document_id`` plus an
ordered tuple of typed segments from the document root down to the cited node
(Part -> section -> subsection -> clause). Structured segments (not a display
string) make the hierarchical relation a clean prefix check and keep the
shared conformance vectors free of any parsing dialect.

The dataclasses are frozen: addresses are values, never mutated in place.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, get_args

#: The node kinds a document tree distinguishes, root-first.
SegmentKind = Literal["document", "part", "section", "subsection", "clause"]

SEGMENT_KINDS: tuple[SegmentKind, ...] = get_args(SegmentKind)


@dataclass(frozen=True, slots=True)
class CitablePathSegment:
    """One step in a citable path: the node's kind and label (e.g. section "49")."""

    kind: SegmentKind
    label: str

    def __post_init__(self) -> None:
        if self.kind not in SEGMENT_KINDS:
            raise ValueError(f"unknown segment kind: {self.kind!r}")
        if not self.label:
            raise ValueError("citable-path segment label must be non-empty")


@dataclass(frozen=True, slots=True)
class CitablePath:
    """The address of one document-tree node: which document, and the ordered
    segments from its root to the node. An empty ``segments`` tuple addresses
    the document root itself."""

    document_id: str
    segments: tuple[CitablePathSegment, ...]

    def __post_init__(self) -> None:
        if not self.document_id:
            raise ValueError("citable path must carry a non-empty document_id")


def parse_citable_path(value: dict) -> CitablePath:
    """Validate and normalize an untyped mapping into a :class:`CitablePath`."""
    if not isinstance(value, dict):
        raise ValueError(f"citable path must be a mapping, got {type(value).__name__}")
    document_id = value.get("documentId")
    if not isinstance(document_id, str) or not document_id:
        raise ValueError("citable path requires a non-empty string documentId")
    raw_segments = value.get("segments", [])
    if not isinstance(raw_segments, list):
        raise ValueError("citable path segments must be a list")
    segments = tuple(
        CitablePathSegment(kind=_require_str(seg, "kind"), label=_require_str(seg, "label"))
        for seg in raw_segments
    )
    return CitablePath(document_id=document_id, segments=segments)


def _require_str(mapping: object, key: str) -> str:
    if not isinstance(mapping, dict) or key not in mapping:
        raise ValueError(f"segment is missing required key {key!r}")
    value = mapping[key]
    if not isinstance(value, str):
        raise ValueError(f"segment {key!r} must be a string")
    return value


def citable_paths_equal(a: CitablePath, b: CitablePath) -> bool:
    """True when two citable paths address the same node of the same document."""
    return a.document_id == b.document_id and a.segments == b.segments


def is_proper_ancestor(maybe_ancestor: CitablePath, maybe_path: CitablePath) -> bool:
    """True when ``maybe_ancestor`` is a strict prefix of ``maybe_path`` within
    the same document. Kinds must agree segment-for-segment, so a mislabeled
    coordinate (a clause "1" vs a subsection "1") is not the same node."""
    if maybe_ancestor.document_id != maybe_path.document_id:
        return False
    if len(maybe_ancestor.segments) >= len(maybe_path.segments):
        return False
    return maybe_ancestor.segments == maybe_path.segments[: len(maybe_ancestor.segments)]


def is_proper_descendant(maybe_descendant: CitablePath, maybe_path: CitablePath) -> bool:
    """True when ``maybe_descendant`` sits strictly lower on the same line."""
    return is_proper_ancestor(maybe_path, maybe_descendant)


def format_citable_path(path: CitablePath) -> str:
    """Render a citable path for human review (e.g. "RTA / Part V / s. 49 / (1)")."""
    parts = [path.document_id]
    for segment in path.segments:
        if segment.kind == "part":
            parts.append(f"Part {segment.label}")
        elif segment.kind == "section":
            parts.append(f"s. {segment.label}")
        elif segment.kind in ("subsection", "clause"):
            parts.append(f"({segment.label})")
        else:  # document
            parts.append(segment.label)
    return " / ".join(parts)


__all__ = [
    "SEGMENT_KINDS",
    "SegmentKind",
    "CitablePath",
    "CitablePathSegment",
    "parse_citable_path",
    "citable_paths_equal",
    "is_proper_ancestor",
    "is_proper_descendant",
    "format_citable_path",
]
